aptos-move/block-executor/src/executor.rs

```diff
@@ -528,38 +528,79 @@ where
+                        },
+                        Err(e) => Err(Error::FallbackToSequential(resource_group_error(format!(
+                            "Error committing resource group {:?}",
+                            e
+                        )))),
+                    }
+                };
+
             let group_metadata_ops = last_input_output.group_metadata_ops(txn_idx);
             let mut finalized_groups = Vec::with_capacity(group_metadata_ops.len());
             let mut maybe_err = None;
             for (group_key, metadata_op) in group_metadata_ops.into_iter() {
                 // finalize_group copies Arc of values and the Tags (TODO: optimize as needed).
-                match versioned_cache
+                let finalized_result = versioned_cache
                     .group_data()
-                    .finalize_group(&group_key, txn_idx)
-                {
+                    .finalize_group(&group_key, txn_idx);
+                match process_finalized_group(finalized_result, metadata_op.is_deletion()) {
                     Ok(finalized_group) => {
-                        // finalize_group already applies the deletions.
-                        if finalized_group.is_empty() != metadata_op.is_deletion() {
-                            maybe_err =
-                                Some(Error::FallbackToSequential(resource_group_error(format!(
-                                "Group is empty = {} but op is deletion = {} in parallel execution",
-                                finalized_group.is_empty(),
-                                metadata_op.is_deletion()
-                            ))));
-                        }
                         finalized_groups.push((group_key, metadata_op, finalized_group));
                     },
-                    Err(e) => {
-                        maybe_err = Some(Error::FallbackToSequential(resource_group_error(
-                            format!("Error committing resource group {:?}", e),
-                        )));
+                    Err(err) => {
+                        maybe_err = Some(err);
+                        break;
                     },
-                };
-
+                }
                 if maybe_err.is_some() {
                     break;
                 }
             }
+
+            if maybe_err.is_none() {
+                if let Some(group_reads_needing_delayed_field_exchange) =
+                    last_input_output.group_reads_needing_delayed_field_exchange(txn_idx)
+                {
+                    for (group_key, metadata_op) in
+                        group_reads_needing_delayed_field_exchange.into_iter()
+                    {
+                        let finalized_result = versioned_cache
+                            .group_data()
+                            .get_last_committed_group(&group_key);
+                        match process_finalized_group(finalized_result, metadata_op.is_deletion()) {
```

Comment:
We can check here that metadata op isn't deletion and finalized_result isn't empty, because this group hasn't been modified, only read, and something was read that required replacing, so it should be impossible to be empty at all (note the difference from the write case).

Reply:
In the process_finalized_group function, I changed the statement to `if finalized_group.is_empty() || metadata_is_deletion { return Err }` Does that work?